File: spk/cli/_cmd_remove.py
from typing import Callable, Any
import argparse
import os
import sys
import termios

import spkrs
import structlog
from ruamel import yaml
from colorama import Fore, Style

import spk
from spk.io import format_ident

from . import _flags

_LOGGER = structlog.get_logger("spk.cli")


def register(
    sub_parsers: argparse._SubParsersAction, **parser_args: Any
) -> argparse.ArgumentParser:

    remove_cmd = sub_parsers.add_parser(
        "remove", aliases=["rm"], help=_remove.__doc__, **parser_args
    )
    remove_cmd.add_argument(
        "--yes", action="store_true", help="Do not ask for confirmations (dangerous!)"
    )
    remove_cmd.add_argument(
        "packages", metavar="PKG", nargs="+", help="The packages to remove"
    )
    _flags.add_repo_flags(remove_cmd, defaults=[])
    remove_cmd.set_defaults(func=_remove)
    return remove_cmd


def _remove(args: argparse.Namespace) -> None:
    """remove a package from a repository."""

    repos = _flags.get_repos_from_repo_flags(args)
    if not repos:
        print(
            f"{Fore.YELLOW}No repositories selected, specify --local-repo (-l) and/or --enable-repo (-r){Fore.RESET}",
            file=sys.stderr,
        )
        sys.exit(1)

    for name in args.packages:

        if "/" not in name and not args.yes:
            answer = input(
                f"{Fore.YELLOW}Are you sure that you want to remove all versions of {name}?{Fore.RESET} [y/N]: "
            )
            if answer.lower() not in ("y", "yes"):
                sys.exit(1)

        for repo_name, repo in repos.items():

            if "/" not in name:
                versions = list(f"{name}/{v}" for v in repo.list_package_versions(name))
            else:
                versions = [name]

            for version in versions:

                ident = spk.api.parse_ident(version)
                if ident.build is not None:
                    _remove_build(repo_name, repo, ident)
                else:
                    _remove_all(repo_name, repo, ident)


def _remove_build(
    repo_name: str, repo: spk.storage.Repository, ident: spk.api.Ident
) -> None:
    try:
        repo.remove_spec(ident)
        _LOGGER.info("removed build spec", pkg=ident, repo=repo_name)
    except spk.storage.PackageNotFoundError:
        _LOGGER.warning("spec not found", pkg=ident, repo=repo_name)
        pass
    try:
        repo.remove_package(ident)
        _LOGGER.info("removed build", pkg=ident, repo=repo_name)
    except spk.storage.PackageNotFoundError:
        _LOGGER.warning("build not found", pkg=ident, repo=repo_name)
        pass


def _remove_all(
    repo_name: str, repo: spk.storage.Repository, ident: spk.api.Ident
) -> None:

    for build in repo.list_package_builds(ident):
        _remove_build(repo_name, repo, build)
    try:
        repo.remove_spec(ident)
        _LOGGER.info("removed spec", pkg=ident, repo=repo_name)
    except spk.storage.PackageNotFoundError:
        _LOGGER.warning("spec not found", pkg=ident, repo=repo_name)
        pass

File: src/cli/cmd_remove.rs
// Copyright (c) 2021 Sony Pictures Imageworks, et al.
// SPDX-License-Identifier: Apache-2.0
// https://github.com/imageworks/spk
use std::io::Write;

use anyhow::Result;
use clap::Args;
use colored::Colorize;
use spk::api;

use super::flags;

/// Remove a package from a repository
#[derive(Args)]
#[clap(visible_alias = "rm")]
pub struct Remove {
    #[clap(flatten)]
    pub repos: flags::Repositories,

    /// Do not ask for confirmations (dangerous!)
    #[clap(short, long)]
    yes: bool,

    #[clap(name = "PKG", required = true)]
    packages: Vec<String>,
}

impl Remove {
    pub fn run(&self) -> Result<i32> {
        let repos = self.repos.get_repos(None)?;
        if repos.is_empty() {
            eprintln!(
                "{}",
                "No repositories selected, specify --local-repo (-l) and/or --enable-repo (-r)"
                    .yellow()
            );
            return Ok(1);
        }

        for name in &self.packages {
            if !name.contains('/') && !self.yes {
                let mut input = String::new();
                print!(
                    "{}",
                    format!("Are you sure that you want to remove all versions of {name}? [y/N]: ")
                        .yellow()
                );
                let _ = std::io::stdout().flush();
                std::io::stdin().read_line(&mut input)?;
                match input.trim() {
                    "y" | "yes" => {}
                    _ => {
                        println!("Removal cancelled");
                        return Ok(1);
                    }
                }
            }

            for (repo_name, repo) in repos.iter() {
                let pkg = api::parse_ident(&name)?;
                let versions = if name.contains('/') {
                    vec![pkg]
                } else {
                    repo.list_package_versions(&name)?
                        .into_iter()
                        .map(|v| pkg.with_version(v))
                        .collect()
                };

                for version in versions {
                    if version.build.is_some() {
                        remove_build(&repo_name, &repo, &version)?;
                    } else {
                        remove_all(&repo_name, &repo, &version)?;
                    }
                }
            }
        }
        Ok(0)
    }
}

fn remove_build(
    repo_name: &String,
    repo: &spk::storage::RepositoryHandle,
    pkg: &spk::api::Ident,
) -> Result<()> {
    let repo_name = repo_name.bold();
    let pretty_pkg = spk::io::format_ident(&pkg);
    match repo.remove_spec(pkg) {
        Ok(()) => {
            tracing::info!("removed build spec {pretty_pkg} from {repo_name}")
        }
        Err(spk::Error::PackageNotFoundError(_)) => {
            tracing::warn!("spec {pretty_pkg} not found in {repo_name}")
        }
        Err(err) => return Err(err.into()),
    }
    match repo.remove_package(pkg) {
        Ok(()) => {
            tracing::info!("removed build      {pretty_pkg} from {repo_name}")
        }
        Err(spk::Error::PackageNotFoundError(_)) => {
            tracing::warn!("build {pretty_pkg} not found in {repo_name}")
        }
        Err(err) => return Err(err.into()),
    }
    Ok(())
}

fn remove_all(
    repo_name: &String,
    repo: &spk::storage::RepositoryHandle,
    pkg: &spk::api::Ident,
) -> Result<()> {
    let pretty_pkg = spk::io::format_ident(pkg);
    for build in repo.list_package_builds(pkg)? {
        remove_build(repo_name, repo, &build)?
    }
    let repo_name = repo_name.bold();
    match repo.remove_spec(pkg) {
        Ok(()) => tracing::info!("removed spec       {pretty_pkg} from {repo_name}"),
        Err(spk::Error::PackageNotFoundError(_)) => {
            tracing::warn!("spec {pretty_pkg} not found in {repo_name}")
        }
        Err(err) => return Err(err.into()),
    }
    Ok(())
}
